include/pybind11/functional.h

```diff
@@ -54,9 +54,20 @@ struct type_caster<std::function<Return(Args...)>> {
             }
         }
 
-        value = [func](Args... args) -> Return {
+        // ensure GIL is held during functor destruction
+        struct func_handle {
+            function f;
+            func_handle(function&& f_) : f(std::move(f_)) {}
+            func_handle(const func_handle&) = default;
+            ~func_handle() {
+                gil_scoped_acquire acq;
+                function kill_f(std::move(f));
```

Comment:
It might be more clear to explicitly reset the function handle and decrease the reference count: `f.release().dec_ref();`

Reply:
Agreed. Done.

Reply:
@davidhewitt causes strange test failure on VS with Python 3.6 on x86 platform, so reverted back to original version.